File: src/lib/vector/chunks.ts
const MAX_CHUNK_CHARS = 1800;

export function splitIntoChunks(text: string, maxChars = MAX_CHUNK_CHARS): string[] {
  const normalized = text.trim();
  if (!normalized) return [];
  if (normalized.length <= maxChars) return [normalized];

  const paragraphs = normalized.split(/\n{2,}/);
  const chunks: string[] = [];
  let current = "";

  for (const paragraph of paragraphs) {
    const piece = paragraph.trim();
    if (!piece) continue;

    if (piece.length > maxChars) {
      if (current) {
        chunks.push(current.trim());
        current = "";
      }
      for (let i = 0; i < piece.length; i += maxChars) {
        chunks.push(piece.slice(i, i + maxChars));
      }
      continue;
    }

    if ((current + "\n\n" + piece).length > maxChars) {
      if (current) chunks.push(current.trim());
      current = piece;
    } else {
      current = current ? `${current}\n\n${piece}` : piece;
    }
  }

  if (current.trim()) chunks.push(current.trim());
  return chunks;
}

export function buildGitHubIndexText(params: {
  repo: string;
  number: number;
  title: string;
  itemType: string;
  aiSummary?: string | null;
  body?: string | null;
  comments?: Array<{ speaker: string; content: string }>;
}): { purpose: string; chunks: string[] } {
  const label = params.itemType === "pull_request" ? "Pull request" : "Issue";
  const purpose =
    params.aiSummary?.trim() ||
    `${label} ${params.repo}#${params.number}: ${params.title}`;

  const header = [
    `Purpose: ${purpose}`,
    `${label}: ${params.repo}#${params.number}`,
    `Title: ${params.title}`,
    params.body?.trim() ? `Description:\n${params.body.trim()}` : null,
    params.aiSummary?.trim() ? `Summary:\n${params.aiSummary.trim()}` : null,
  ]
    .filter(Boolean)
    .join("\n\n");

  const commentBlock =
    params.comments && params.comments.length > 0
      ? `Comments:\n${params.comments
          .slice(-12)
          .map((c) => `${c.speaker}: ${c.content}`)
          .join("\n")}`
      : "";

  const full = [header, commentBlock].filter(Boolean).join("\n\n");
  return { purpose, chunks: splitIntoChunks(full) };
}

export function buildMeetingIndexText(params: {
  title?: string | null;
  aiSummary: string;
  decisions?: string[];
  actionItems?: Array<{ text: string }>;
}): { purpose: string; chunks: string[] } {
  const purpose = params.aiSummary.split("\n")[0]?.slice(0, 240) || params.title || "Meeting notes";
  const body = [
    params.title ? `Meeting: ${params.title}` : null,
    `Summary:\n${params.aiSummary}`,
    params.decisions?.length
      ? `Decisions:\n${params.decisions.map((d) => `- ${d}`).join("\n")}`
      : null,
    params.actionItems?.length
      ? `Action items:\n${params.actionItems.map((a) => `- ${a.text}`).join("\n")}`
      : null,
  ]
    .filter(Boolean)
    .join("\n\n");

  return { purpose, chunks: splitIntoChunks(body) };
}

File: backend/app/services/vector/chunks.py
MAX_CHUNK_CHARS = 1800


def split_into_chunks(text: str, max_chars: int = MAX_CHUNK_CHARS) -> list[str]:
    normalized = text.strip()
    if not normalized:
        return []
    if len(normalized) <= max_chars:
        return [normalized]

    paragraphs = normalized.split("\n\n")
    chunks: list[str] = []
    current = ""

    for paragraph in paragraphs:
        piece = paragraph.strip()
        if not piece:
            continue

        if len(piece) > max_chars:
            if current:
                chunks.append(current.strip())
                current = ""
            for i in range(0, len(piece), max_chars):
                chunks.append(piece[i : i + max_chars])
            continue

        combined = f"{current}\n\n{piece}" if current else piece
        if len(combined) > max_chars:
            if current:
                chunks.append(current.strip())
            current = piece
        else:
            current = combined

    if current.strip():
        chunks.append(current.strip())
    return chunks


def build_github_index_text(
    repo: str,
    number: int,
    title: str,
    item_type: str,
    ai_summary: str | None = None,
    body: str | None = None,
    comments: list[dict[str, str]] | None = None,
) -> dict[str, object]:
    label = "Pull request" if item_type == "pull_request" else "Issue"
    purpose = (
        ai_summary.strip()
        if ai_summary and ai_summary.strip()
        else f"{label} {repo}#{number}: {title}"
    )

    header_parts = [
        f"Purpose: {purpose}",
        f"{label}: {repo}#{number}",
        f"Title: {title}",
        f"Description:\n{body.strip()}" if body and body.strip() else None,
        f"Summary:\n{ai_summary.strip()}" if ai_summary and ai_summary.strip() else None,
    ]
    header = "\n\n".join(p for p in header_parts if p)

    comment_block = ""
    if comments:
        recent = comments[-12:]
        comment_lines = "\n".join(f"{c['speaker']}: {c['content']}" for c in recent)
        comment_block = f"Comments:\n{comment_lines}"

    full = "\n\n".join(p for p in [header, comment_block] if p)
    return {"purpose": purpose, "chunks": split_into_chunks(full)}


def build_meeting_index_text(
    ai_summary: str,
    title: str | None = None,
    decisions: list[str] | None = None,
    action_items: list[dict[str, str]] | None = None,
) -> dict[str, object]:
    purpose = (
        ai_summary.split("\n")[0][:240]
        if ai_summary
        else (title or "Meeting notes")
    )
    body_parts = [
        f"Meeting: {title}" if title else None,
        f"Summary:\n{ai_summary}",
        f"Decisions:\n" + "\n".join(f"- {d}" for d in decisions) if decisions else None,
        (
            f"Action items:\n" + "\n".join(f"- {a['text']}" for a in action_items)
            if action_items
            else None
        ),
    ]
    body = "\n\n".join(p for p in body_parts if p)
    return {"purpose": purpose, "chunks": split_into_chunks(body)}
